### ISSUE TITLE
virt-operator pod crash becasue of securityContext.seccompProfile: type: RuntimeDefault

### ISSUE THREAD

**Issue by zhhray** [kubevirt/kubevirt#9540]
**What happened**:
virt-operator pod crash when deploy kubevirt operator in my k8s cluster.

**What you expected to happen**:
virt-operator pod do not crash

**How to reproduce it (as minimally and precisely as possible)**:
It's an occasional problem that can occur in particular circumstances.

**Additional context**:

**Environment**:
- KubeVirt version (use `virtctl version`): v0.59.0
- Kubernetes version (use `kubectl version`): v1.23.16
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): 
```sh
cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`): 
```sh
uname  -a
Linux localhost.localdomain 3.10.0-1160.el7.x86_64 #1 SMP Mon Oct 19 16:18:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools: kubevirt operator
- Others: N/A

**Comment by zhhray** [comment]
virt-operator pod logs:
```sh
runtime/cgo: pthread_create failed: Operation not permitted
SIGABRT: abort
PC=0x19733ac m=0 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: g 0: unknown pc 0x19733ac
stack: frame={sp:0x7ffd1dbf0e40, fp:0x0} stack=[0x7ffd1d3f23c0,0x7ffd1dbf13d0)
0x00007ffd1dbf0d40:  0x000000000041a2d9 <runtime.sysMapOS+0x0000000000000039>  0x0000000000000000 
0x00007ffd1dbf0d50:  0x00007f777da85000  0x0000000000405240 <runtime.mmap.func1+0x0000000000000000> 
0x00007ffd1dbf0d60:  0x0000000000000000  0x0000000000000000 
0x00007ffd1dbf0d70:  0x0000000000407000 <runtime.cgoCheckUsingType+0x0000000000000080>  0x00007f777fc96000 
0x00007ffd1dbf0d80:  0x00007ffd1dbf0db0  0x00000000018fe6bc 
0x00007ffd1dbf0d90:  0x00007f777fc96000  0x000000000046c95e <runtime.callCgoMmap+0x000000000000003e> 
0x00007ffd1dbf0da0:  0x00007ffd1dbf0da8  0x00007f777fc96000 
0x00007ffd1dbf0db0:  0x00007ffd1dbf0e20  0x00000000018fe6bc 
0x00007ffd1dbf0dc0:  0x000000000040528f <runtime.mmap.func1+0x000000000000004f>  0x000000000046c95e <runtime.callCgoMmap+0x000000000000003e> 
0x00007ffd1dbf0dd0:  0x00007ffd1dbf0dd8  0x0000000000000007 
0x00007ffd1dbf0de0:  0x00007ffd1dbf0ea0  0x000000000042c5cd <runtime.(*pageAlloc).update+0x00000000000003ad> 
0x00007ffd1dbf0df0:  0x00007f77a4d16c00  0x00000000018fe6bc 
0x00007ffd1dbf0e00:  0x0000000000010000  0x000000000046c95e <runtime.callCgoMmap+0x000000000000003e> 
0x00007ffd1dbf0e10:  0x00000000019726f0  0x000000001c000004 
0x00007ffd1dbf0e20:  0x00000000019014b0  0x0000000000000000 
0x00007ffd1dbf0e30:  0x00007f777dc84fff  0x000000000197339e 
0x00007ffd1dbf0e40: <0x0000000000000400  0x0000000000000000 
0x00007ffd1dbf0e50:  0x0000000000000010  0x0000000000000000 
0x00007ffd1dbf0e60:  0x0000000000000120  0x0000000000000110 
0x00007ffd1dbf0e70:  0x00000000004299e5 <runtime.(*mheap).alloc+0x0000000000000065>  0x0000000000000012 
0x00007ffd1dbf0e80:  0x0000000000000400  0x0000000001939e11 
0x00007ffd1dbf0e90:  0x00007f777d964640  0x00007ffd1dbf1150 
0x00007ffd1dbf0ea0:  0x00007ffd1dbf0fbe  0x00007ffd1dbf0fbf 
0x00007ffd1dbf0eb0:  0x00007f777d964640  0x000000000190bcc5 
0x00007ffd1dbf0ec0:  0x000000000042fc65 <runtime.(*pallocData).allocRange+0x0000000000000045>  0x5b84c7eea5a5c800 
0x00007ffd1dbf0ed0:  0x00000000038833c0  0x0000000000000006 
0x00007ffd1dbf0ee0:  0x0000000003884650  0x0000000000000178 
0x00007ffd1dbf0ef0:  0x00000000034a5fe0  0x0000000001963d16 
0x00007ffd1dbf0f00:  0x00000000034d7020  0x00000000004011d8 
0x00007ffd1dbf0f10:  0x0000000000000020  0x0000000000000000 
0x00007ffd1dbf0f20:  0x0000000000000000  0x5b84c7eea5a5c800 
0x00007ffd1dbf0f30:  0x00007ffd1dbf1000  0x00007f777d964640 
runtime: g 0: unknown pc 0x19733ac
stack: frame={sp:0x7ffd1dbf0e40, fp:0x0} stack=[0x7ffd1d3f23c0,0x7ffd1dbf13d0)
0x00007ffd1dbf0d40:  0x000000000041a2d9 <runtime.sysMapOS+0x0000000000000039>  0x0000000000000000 
0x00007ffd1dbf0d50:  0x00007f777da85000  0x0000000000405240 <runtime.mmap.func1+0x0000000000000000> 
0x00007ffd1dbf0d60:  0x0000000000000000  0x0000000000000000 
0x00007ffd1dbf0d70:  0x0000000000407000 <runtime.cgoCheckUsingType+0x0000000000000080>  0x00007f777fc96000 
0x00007ffd1dbf0d80:  0x00007ffd1dbf0db0  0x00000000018fe6bc 
0x00007ffd1dbf0d90:  0x00007f777fc96000  0x000000000046c95e <runtime.callCgoMmap+0x000000000000003e> 
0x00007ffd1dbf0da0:  0x00007ffd1dbf0da8  0x00007f777fc96000 
0x00007ffd1dbf0db0:  0x00007ffd1dbf0e20  0x00000000018fe6bc 
0x00007ffd1dbf0dc0:  0x000000000040528f <runtime.mmap.func1+0x000000000000004f>  0x000000000046c95e <runtime.callCgoMmap+0x000000000000003e> 
0x00007ffd1dbf0dd0:  0x00007ffd1dbf0dd8  0x0000000000000007 
0x00007ffd1dbf0de0:  0x00007ffd1dbf0ea0  0x000000000042c5cd <runtime.(*pageAlloc).update+0x00000000000003ad> 
0x00007ffd1dbf0df0:  0x00007f77a4d16c00  0x00000000018fe6bc 
0x00007ffd1dbf0e00:  0x0000000000010000  0x000000000046c95e <runtime.callCgoMmap+0x000000000000003e> 
0x00007ffd1dbf0e10:  0x00000000019726f0  0x000000001c000004 
0x00007ffd1dbf0e20:  0x00000000019014b0  0x0000000000000000 
0x00007ffd1dbf0e30:  0x00007f777dc84fff  0x000000000197339e 
0x00007ffd1dbf0e40: <0x0000000000000400  0x0000000000000000 
0x00007ffd1dbf0e50:  0x0000000000000010  0x0000000000000000 
0x00007ffd1dbf0e60:  0x0000000000000120  0x0000000000000110 
0x00007ffd1dbf0e70:  0x00000000004299e5 <runtime.(*mheap).alloc+0x0000000000000065>  0x0000000000000012 
0x00007ffd1dbf0e80:  0x0000000000000400  0x0000000001939e11 
0x00007ffd1dbf0e90:  0x00007f777d964640  0x00007ffd1dbf1150 
0x00007ffd1dbf0ea0:  0x00007ffd1dbf0fbe  0x00007ffd1dbf0fbf 
0x00007ffd1dbf0eb0:  0x00007f777d964640  0x000000000190bcc5 
0x00007ffd1dbf0ec0:  0x000000000042fc65 <runtime.(*pallocData).allocRange+0x0000000000000045>  0x5b84c7eea5a5c800 
0x00007ffd1dbf0ed0:  0x00000000038833c0  0x0000000000000006 
0x00007ffd1dbf0ee0:  0x0000000003884650  0x0000000000000178 
0x00007ffd1dbf0ef0:  0x00000000034a5fe0  0x0000000001963d16 
0x00007ffd1dbf0f00:  0x00000000034d7020  0x00000000004011d8 
0x00007ffd1dbf0f10:  0x0000000000000020  0x0000000000000000 
0x00007ffd1dbf0f20:  0x0000000000000000  0x5b84c7eea5a5c800 
0x00007ffd1dbf0f30:  0x00007ffd1dbf1000  0x00007f777d964640 

goroutine 1 [running]:
runtime.systemstack_switch()
        GOROOT/src/runtime/asm_amd64.s:459 fp=0xc0000da780 sp=0xc0000da778 pc=0x468940
runtime.main()
        GOROOT/src/runtime/proc.go:170 +0x6d fp=0xc0000da7e0 sp=0xc0000da780 pc=0x43bb2d
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc0000da7e8 sp=0xc0000da7e0 pc=0x46ab61

rax    0x0
rbx    0x38833c0
rcx    0xffffffffffffffff
rdx    0x6
rdi    0x1
rsi    0x1
rbp    0x1
rsp    0x7ffd1dbf0e40
r8     0x7ffd1dbf0f10
r9     0x3365960
r10    0x8
r11    0x246
r12    0x6
r13    0x178
r14    0x34a5fe0
r15    0x7f777da8535b
rip    0x19733ac
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
```

**Comment by zhhray** [comment]
If I manually remove the securityContext.seccompProfile related configuration from deployment, the problem will disappear.
Part of virt-operator yaml follows:
```yaml
       securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          seccompProfile:
            type: RuntimeDefault
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/virt-operator/certificates
          name: kubevirt-operator-certs
          readOnly: true
        - mountPath: /profile-data
          name: profile-data
      dnsPolicy: ClusterFirst
      priorityClassName: kubevirt-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        runAsNonRoot: true
        seccompProfile:
          type: RuntimeDefault
```
after manual modification, yaml like this:
```yaml
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/virt-operator/certificates
          name: kubevirt-operator-certs
          readOnly: true
        - mountPath: /profile-data
          name: profile-data
      dnsPolicy: ClusterFirst
      priorityClassName: kubevirt-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        runAsNonRoot: true
```

**Comment by zhhray** [comment]
> It seems that virt-operator executes one of the syscalls not allowed in the default seccomp profile. Can you please have a look at the auditing logs? The error could be reported also on the pod description, could you please copy-paste the output of `kubectl describe po -n kubevirt <virt-operator-pod-name>`

I'm sorry that I can't provide it.
This problem disappears by upgrading the docker version.
docker version from `v20.10.7` upgrade to `v20.10.23-2`.
Therefore, whether to use  `securityContext. seccompProfile.type:  RuntimeDefault` is a risky configuration item, which may cause exceptions on different versions of CRI?

**Comment by vasiliy-ul** [comment]
This reminds me of an old issue https://github.com/moby/moby/issues/42680, when glibc started using clone3 in pthread_create:

> The latest glibc now attempts to use 'clone3' by default. For backwards compatibility it will look for ENOSYS errno and fallback to "clone". The EPERM errno meanwhile is treated as a fatal error.
The default seccomp filter installed by docker is causing EPERM and so this breaks the glibc fallback.

In this case, we have a recent code from virt-operator that runs on a relatively old host with centos 7. If the host uses an old version of the container rintime and/or libseccomp then chances are that it might be the same issue....

**Comment by alicefr** [comment]
> I'm sorry that I can't provide it. This problem disappears by upgrading the docker version. docker version from `v20.10.7` upgrade to `v20.10.23-2`. Therefore, whether to use `securityContext.seccompProfile.type: RuntimeDefault` is a risky configuration item, which may cause exceptions on different versions of CRI?

Yes, the default seccomp profile depends on the container runtime your cluster is using. Great, probably this was fixed in  a more recent docker version and probably the reason is what @vasiliy-ul mentioned above.

**Comment by xpivarc** [comment]
@vasiliy-ul +1
I would recommend checking out what is the version and what runtime is used. I think we can then setup our runtime default if upgrade would not be feasible

**Comment by vasiliy-ul** [comment]
For the record: this has been fixed in docker since `v20.10.10` https://github.com/moby/moby/releases/tag/v20.10.10
 > Add support for clone3 syscall in the default seccomp policy to support running
containers based on recent versions of Fedora and Ubuntu. https://github.com/moby/moby/pull/42836.

**Comment by xpivarc** [comment]
/close

You can use a custom seccomp with https://github.com/kubevirt/kubevirt/pull/8917, available in v0.59+ or upgrade docker as suggested.

Please reopen if the issue persist
